### PR TITLE
Phonewords to Phone Numbers

### PR DESCRIPTION
**Pull request by moollaza**
Continuation of #849 

/cc @jagtalon @MrChrisW @stevelippert 

2015-01-29_15h37_36

### CODE REVIEW

**Comment by moollaza** [comment]
My only gripe about this is that it doesn't properly format the output number. It preserver the original formatting. I attempted to normalize the numbers, but it was going to be a decent amount of work (lots of edge cases and intl. formats) so I've skipped it for now. Can ask the community to tackle that. This could possible evolve into a "phone number formatter" IA if we want.

File: lib/DDG/Goodie/PhoneAlphabet.pm [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            twitter => ["https://twitter.com/stevelippert", "stevelippert"];
+
+triggers any => 'to digit', 'to digits', 'to phone', 'to phone number';
+
+handle remainder => sub {
+    # Return unless it looks like a phone number
+    return unless ($_ =~ /[-0-9A-Za-z]{6,15}$/);
+    # Return if it's a Base64 number
+    return if ($_ =~ /^0x\d+$/);
+    return unless $_;
```
Comment (mintsoft) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: I think this is unnecessary? doesn't L23 handle this?

File: lib/DDG/Goodie/PhoneAlphabet.pm [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    # Lower case everything.
+    my $num = lc $_;
+    # Use a regex to replace each letter with the corresponding number from the phone key pad.
+    $num =~ tr/abcdefghijklmnopqrstuvwxyz/22233344455566677778889999/;
+
+    return "Phone Number: $num",
+    structured_answer => {
+        input     => ["$_"],
```
Comment (mintsoft) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: Is this html_enc'd by duckpan?

File: lib/DDG/Goodie/PhoneAlphabet.pm [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+code_url "https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/PhoneAlphabet.pm";
+attribution github => ["https://github.com/stevelippert", "Steve Lippert"],
+            twitter => ["https://twitter.com/stevelippert", "stevelippert"];
+
+triggers any => 'to digit', 'to digits', 'to phone', 'to phone number';
+
+handle remainder => sub {
+    # Return unless it looks like a phone number
+    return unless ($_ =~ /[-0-9A-Za-z]{6,15}$/);
+    # Return if it's a Base64 number
```
Comment (mintsoft) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: Hex number surely?

**Comment by mintsoft** [comment]
@moollaza I was going to pick this up this weekend but you've beaten me to the punch!

File: lib/DDG/Goodie/PhoneAlphabet.pm (hunk 3 above)
Comment (moollaza) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: Haha, definitely didn't sleep enough last night. You're right!

File: lib/DDG/Goodie/PhoneAlphabet.pm (hunk 2 above)
Comment (moollaza) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: Good catch, I'm not actually sure, but I think it is

**Comment by moollaza** [comment]
@mintsoft if you want to take it further you're welcome to. I'll fixup the stuff you mentioned, but you can pick it up after if you'd like.

File: lib/DDG/Goodie/PhoneAlphabet.pm (hunk 2 above)
Comment (moollaza) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: Turns out it isn't. I've added a commit to encode it.

Comment (moollaza) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: I wonder if we should add a param to force encoding:

```perl
structured_answer => {
    encode_output => 1,
    input => ["..."],
    operation => "...",
    answer => "..."
}
```

**Comment by jagtalon** [comment]
@moollaza nice! @abeyang https://ddh1.duckduckgo.com/?q=1-800-GOT-PUNK+to+digits&ia=answer

**Comment by abeyang** [comment]
@moollaza @jagtalon looks good. Remind me -- we can't just get rid of the "Phone number:" in the byline?

**Comment by MrChrisW** [comment]
@abeyang We can remove the `operation` value that returns "Phone number" quite easily I think.

**Comment by jagtalon** [comment]
@abeyang so you'd rather have them removed?

**Comment by moollaza** [comment]
@abeyang I like having an operation so it's clear how we interpreted your query and what it was we did. Would a different `operation` make more sense? Maybe "Digitize" or "Convert"?

**Comment by chrismorast** [comment]
I think it should stay in the byline.

**Comment by moollaza** [comment]
Alright then, looks like this is good to merge. Thanks @mintsoft @jagtalon @abeyang @chrismorast @MrChrisW

File: lib/DDG/Goodie/PhoneAlphabet.pm (hunk 2 above)
Comment (mintsoft) [inline lib/DDG/Goodie/PhoneAlphabet.pm]: I think we probably should do something similiar, however we need to balance XSS against HTML output for each goodie somehow

**Comment by moollaza** [comment]
@MrChrisW @stevelippert congrats! It's live: https://duckduckgo.com/?q=1-800-GOT-JUNK+to+digits&ia=answer

**Pull request merged**